File: days/2/puzzles/1/program.js
let input = '';

process.stdin.on('data', chunk => input += chunk);

process.stdin.on('end', () => {
  const program = input.trim().split(',');
  console.log(recoverFrom1202Alarm(program));
});

function recoverFrom1202Alarm(program) {
  program[1] = '12';
  program[2] = '2';

  return runProgram(program);
}

function runProgram(program) {
  for (let i = 0; i < program.length; i += 4) {
    const opCode = parseInt(program[i], 10);
    const input1Position = parseInt(program[i + 1], 10);
    const input2Position = parseInt(program[i + 2], 10);
    const resultPosition = parseInt(program[i + 3], 10);
    const input1 = parseInt(program[input1Position], 10);
    const input2 = parseInt(program[input2Position], 10);

    let result;

    if (opCode === 1) {
      result = input1 + input2;
    } else if (opCode === 2) {
      result = input1 * input2;
    } else if (opCode === 99) {
      return program[0];
    } else {
      throw new Error(`Unrecognized Op Code: ${opCode} at position: ${i}`);
    }

    if (!isNaN(resultPosition) && !isNaN(result)) {
      program[resultPosition] = result + '';
    } else {
      throw new Error(`Invalid inputs for Op Code at position ${i}.`);
    }
  }

  throw new Error('End of input reached with no halting instruction.');
}

File: days/2/puzzles/1/program.py
import sys

memory = sys.stdin.read().split(',')

memory[1] = '12'
memory[2] = '2'

i = 0

while i < len(memory):
  op_code = int(memory[i])
  input1_pos = int(memory[i+1])
  input2_pos = int(memory[i+2])
  result_pos = int(memory[i+3])
  input1 = int(memory[input1_pos])
  input2 = int(memory[input2_pos])

  if op_code == 1:
    memory[result_pos] = str(input1 + input2)
  elif op_code == 2:
    memory[result_pos] = str(input1 * input2)
  else:
    break

  i += 4

print(memory[0])
